[Form] Support \DateTimeImmutable

vudaltsov opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #9508
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/8920

This PR implements `input=datetime_immutable`. Replaces #25273.

vudaltsov committed:
Added input=datetime_immutable for DateType, TimeType, DateTimeType.

jvasseur commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToDateTimeTransformer.php:
```diff
@@ -0,0 +1,67 @@
+            throw new TransformationFailedException('Expected a \DateTimeImmutable.');
+        }
+
+        return new \DateTime($value->format(\DateTime::RFC3339));
```
Maybe use `createFromFormat` here to remove the need for PHP to guess the date format.

vudaltsov replied:
@jvasseur , good idea, thanx.

vudaltsov committed:
Use createFromFormat when transforming DateTimeImmutable -> DateTime.

hhamon commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToDateTimeTransformer.php:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @throws TransformationFailedException If the given value is not a \DateTimeImmutable
+     */
+    public function transform($value): ?\DateTime
```
Adding the return type hint breaks the prototype.

hhamon commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToDateTimeTransformer.php:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @throws TransformationFailedException If the given value is not a \DateTime
+     */
+    public function reverseTransform($value): ?\DateTimeImmutable
```
Adding the return type hint breaks the prototype.

vudaltsov replied:
thank you, fixed!

nicolas-grekas commented:
Status: needs work

vudaltsov committed:
Removed return typehints in DateTimeImmutableToDateTimeTransformer.

vudaltsov commented:
Status: Needs Review

vudaltsov committed:
Marked DateTimeImmutableToDateTimeTransformer as final.

vudaltsov commented:
Anything else needed here?

HeahDude reviewed (approved):
This is a nice approach to leverage the feature without too much overhead, until we make the required change to make it the default norm format.
LGTM

nicolas-grekas reviewed (approved):
(small rebase needed)

vudaltsov committed:
Use assertEquals to compare \DateTimeImmutable.

vudaltsov committed:
Added line to CHANGELOG.

vudaltsov commented:
Rebased. Added a changelog entry and replaced previously added `assertDateTimeImmutableEquals` with `assertEqual`.

If everything's good after this, I will squash commits.

vudaltsov committed:
Added instanceof assertions.

fabpot commented:
Thank you @vudaltsov.

ste93cry commented:
Are there plans to port this to at least Symfony 3.4?

derrabus commented:
@ste93cry Since this is considered to be a new feature, it most likely won't land in 3.4, which is in bugfix-only mode.

vudaltsov commented:
@ste93cry , during this week-end I will create a small bundle to add this functionality to earlier Symfony versions.

vudaltsov commented:
Created a [polyfill bundle](https://github.com/ruvents/ruwork-polyfill-form-dti-bundle) for those who want to use `input=datetime_immutable` with Symfony (`>=2.8 <4.1`).
